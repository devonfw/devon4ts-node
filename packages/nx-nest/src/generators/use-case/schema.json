{
  "$schema": "https://json-schema.org/schema",
  "$id": "UseCase",
  "title": "",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use?"
    },
    "nameAndDirectoryFormat": {
      "description": "Whether to generate the component in the directory as provided, relative to the current working directory and ignoring the project (`as-provided`) or generate it using the project and directory relative to the workspace root (`derived`).",
      "type": "string",
      "enum": ["as-provided", "derived"]
    },
    "directory": {
      "description": "Directory where the generated files are placed. when `--nameAndDirectoryFormat=as-provided`, it will be relative to the current working directory. otherwise, it will be relative to the workspace root.",
      "type": "string",
      "aliases": ["dir", "path"]
    },
    "specSameFolder": {
      "description": "Generate spec file in the same folder as use-case implementation.",
      "type": "boolean",
      "default": false
    },
    "skipFormat": {
      "description": "Skip formatting files.",
      "type": "boolean",
      "default": false,
      "x-priority": "internal"
    }
  },
  "additionalProperties": false,
  "required": ["name"]
}
