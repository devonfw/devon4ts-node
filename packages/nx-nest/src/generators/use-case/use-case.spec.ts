import { Tree } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { moduleGenerator } from '@nx/nest';
import { ApplicationGeneratorOptions } from '@nx/nest/src/generators/application/schema';
import applicationGenerator from '../application/application';
import useCaseGenerator from './use-case';

describe('use-case generator', () => {
  let tree: Tree;
  const applicationOptions: ApplicationGeneratorOptions = {
    name: 'test',
    directory: 'apps/test',
    skipFormat: true,
    skipPackageJson: true,
  };

  beforeEach(async () => {
    tree = createTreeWithEmptyWorkspace();
    await applicationGenerator(tree, applicationOptions);
  });

  it('should generate the use-case in the application layer and add it to app.module', async () => {
    await useCaseGenerator(tree, { name: 'generate-test', path: 'apps/test/src/app' });
    expect(tree.exists('apps/test/src/app/application/generate-test.use-case.ts')).toBeTruthy();
    expect(tree.read('apps/test/src/app/application/generate-test.use-case.ts')?.toString('utf-8')).toMatchSnapshot();
    expect(tree.read('apps/test/src/app/app.module.ts')?.toString('utf-8')).toMatchSnapshot();
  });

  it('should generate the use-case in the application layer of the specified module', async () => {
    await moduleGenerator(tree, {
      module: 'test-module',
      path: 'apps/test/src/app/test-module/test-module',
    });
    await useCaseGenerator(tree, { name: 'generate-test', path: 'apps/test/src/app/test-module' });

    expect(tree.exists('apps/test/src/app/test-module/application/generate-test.use-case.ts')).toBeTruthy();
    expect(tree.exists('apps/test/src/app/application/generate-test.use-case.ts')).toBeFalsy();
    expect(tree.exists('apps/test/src/app/test-module/application/generate-test.use-case.test.ts')).toBeFalsy();
    expect(tree.exists('apps/test/src/test/test-module/application/generate-test.use-case.test.ts')).toBeTruthy();
    expect(
      tree.read('apps/test/src/app/test-module/application/generate-test.use-case.ts')?.toString('utf-8'),
    ).toMatchSnapshot();
    expect(
      tree.read('apps/test/src/test/test-module/application/generate-test.use-case.test.ts')?.toString('utf-8'),
    ).toMatchSnapshot();
    expect(tree.read('apps/test/src/app/test-module/test-module.module.ts')?.toString('utf-8')).toMatchSnapshot();
  });

  it('should generate the specs in the same folder if specSameFolder is true', async () => {
    await moduleGenerator(tree, {
      module: 'test-module',
      path: 'apps/test/src/app/test-module/test-module',
    });
    await useCaseGenerator(tree, {
      name: 'generate-test',
      path: 'apps/test/src/app/test-module',
      specSameFolder: true,
    });
    expect(tree.exists('apps/test/src/app/test-module/application/generate-test.use-case.ts')).toBeTruthy();
    expect(tree.exists('apps/test/src/app/application/generate-test.use-case.ts')).toBeFalsy();
    expect(tree.exists('apps/test/src/app/test-module/application/generate-test.use-case.test.ts')).toBeTruthy();
    expect(tree.exists('apps/test/src/test/test-module/application/generate-test.use-case.test.ts')).toBeFalsy();
    expect(
      tree.read('apps/test/src/app/test-module/application/generate-test.use-case.ts')?.toString('utf-8'),
    ).toMatchSnapshot();
    expect(
      tree.read('apps/test/src/app/test-module/application/generate-test.use-case.test.ts')?.toString('utf-8'),
    ).toMatchSnapshot();
    expect(tree.read('apps/test/src/app/test-module/test-module.module.ts')?.toString('utf-8')).toMatchSnapshot();
  });
});
