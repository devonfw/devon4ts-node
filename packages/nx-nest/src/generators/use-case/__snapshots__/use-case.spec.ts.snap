// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`use-case generator should generate the specs in the same folder if specSameFolder is true 1`] = `
"import { Injectable } from '@nestjs/common';

@Injectable()
export class GenerateTest {
  execute(): void {
    throw new Error('Not implemented!');
  }
}
"
`;

exports[`use-case generator should generate the specs in the same folder if specSameFolder is true 2`] = `
"import { Test, TestingModule } from '@nestjs/testing';
import { GenerateTest } from '/generate-test.use-case';

describe('GenerateTest', () => {
  let useCase: GenerateTest;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [GenerateTest],
    }).compile();

    useCase = module.get<GenerateTest>(GenerateTest);
  });

  it('should be defined', () => {
    expect(useCase).toBeDefined();
  });
});
"
`;

exports[`use-case generator should generate the specs in the same folder if specSameFolder is true 3`] = `
"import { Module } from '@nestjs/common';
import { GenerateTest } from './application/generate-test.use-case';

@Module({
  providers: [GenerateTest],
})
export class TestModuleModule {}
"
`;

exports[`use-case generator should generate the use-case in the application layer and add it to app.module 1`] = `
"import { Injectable } from '@nestjs/common';

@Injectable()
export class GenerateTest {
  execute(): void {
    throw new Error('Not implemented!');
  }
}
"
`;

exports[`use-case generator should generate the use-case in the application layer and add it to app.module 2`] = `
"import { Module } from '@nestjs/common';
import { CoreModule } from './core/core.module';
import { GenerateTest } from './application/generate-test.use-case';

@Module({
  imports: [CoreModule],
  controllers: [],
  providers: [GenerateTest],
})
export class AppModule {}
"
`;

exports[`use-case generator should generate the use-case in the application layer of the specified module 1`] = `
"import { Injectable } from '@nestjs/common';

@Injectable()
export class GenerateTest {
  execute(): void {
    throw new Error('Not implemented!');
  }
}
"
`;

exports[`use-case generator should generate the use-case in the application layer of the specified module 2`] = `
"import { Test, TestingModule } from '@nestjs/testing';
import { GenerateTest } from '../../../app/test-module/application/generate-test.use-case';

describe('GenerateTest', () => {
  let useCase: GenerateTest;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [GenerateTest],
    }).compile();

    useCase = module.get<GenerateTest>(GenerateTest);
  });

  it('should be defined', () => {
    expect(useCase).toBeDefined();
  });
});
"
`;

exports[`use-case generator should generate the use-case in the application layer of the specified module 3`] = `
"import { Module } from '@nestjs/common';
import { GenerateTest } from './application/generate-test.use-case';

@Module({
  providers: [GenerateTest],
})
export class TestModuleModule {}
"
`;
